src/swap/swap_lib_calls.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#define RUN_APPLICATION 1
+
+#define SIGN_TRANSACTION 2
+
+#define CHECK_ADDRESS 3
+
+#define GET_PRINTABLE_AMOUNT 4
+
+/*
+ * Amounts are stored as bytes, with a max size of 16 (see protobuf
+ * specifications). Max 16B integer is 340282366920938463463374607431768211455
+ * in decimal, which is a 32-long char string.
+ * The printable amount also contains spaces, the ticker symbol (with variable
+ * size, up to 12 in Ethereum for instance) and a terminating null byte, so 50
+ * bytes total should be a fair maximum.
+ */
+#define MAX_PRINTABLE_AMOUNT_SIZE 50
+
+// structure that should be send to specific coin application to get address
+typedef struct check_address_parameters_s {
+    // IN
+    unsigned char *coin_configuration;
+    unsigned char coin_configuration_length;
+    // serialized path, segwit, version prefix, hash used, dictionary etc.
+    // fields and serialization format depends on specific coin app
+    unsigned char *address_parameters;
+    unsigned char address_parameters_length;
+    char *address_to_check;
+    char *extra_id_to_check;
+    // OUT
+    int result;
+} check_address_parameters_t;
+
+// structure that should be send to specific coin application to get printable amount
+typedef struct get_printable_amount_parameters_s {
+    // IN
+    unsigned char *coin_configuration;
+    unsigned char coin_configuration_length;
+    unsigned char *amount;
+    unsigned char amount_length;
+    bool is_fee;
+    // OUT
+    char printable_amount[MAX_PRINTABLE_AMOUNT_SIZE];
+} get_printable_amount_parameters_t;
+
+typedef struct create_transaction_parameters_s {
+    unsigned char *coin_configuration;
+    unsigned char coin_configuration_length;
+    unsigned char *amount;
```

Comment:
is it possible to keep track of both - spending amount and receiving amount? I think when user confirms swap on the ledger they need to approve both

Reply:
When the user confirms the swap on his app he will indeed need to approve both the spending and the receiving amount

The spending amount is given a first time to the app of the OUT coin for formatting and format verification ( get_printable_amount() )
It is then given a second time to the app of the OUT coin for the SIGN_TRANSACTION call

The receiving amount is a to the app of the IN coin for formatting and format verification ( get_printable_amount() )
It is not given to the app of the OUT coin app because it would not know what to do with the value
(heavily depends on the currency, some use exponents, some encoded values, etc)

For example if we do sol -> eth, we can't give eth formatted value to the sol app